### ISSUE TITLE
Custom Column expression `substring()` returns nothing for MySQL

### ISSUE THREAD

**Issue by flamber** [metabase/metabase#12445]
**Describe the bug**
Using the expression function `substring()` when querying MySQL returns empty (`""`) column.

**To Reproduce**
1. Custom question > select MySQL/MariaDB database and a table with string column
2. Click on Custom Column > formula: `substring([string_column], 0, 5)` name: "test"
3. Visualize > shows empty values in the "test" column.

**Expected behavior**
Expect the snippet returned ;-)

**Information about your Metabase Installation:**
Metabase 0.35.3 with multiple backends and internal instance.

**Additional context**
Postgres and H2 seems to work (haven't tested other databases)

**Comment by camsaul** [comment]
As part of this we should write a test to make sure `substring` support works for all drivers we support it for

**Comment by camsaul** [comment]
So, we actually do have a test for `substring` here:

https://github.com/metabase/metabase/blob/master/test/metabase/query_processor_test/string_extracts_test.clj#L43

So we should try to repro this using the steps flamber posted and then try that specific query. Maybe the query looks different than what we expect or maybe we're not testing a certain variation of the clause

**Comment by robdaemon** [comment]
This is a MySQL-ism - per the MySQL docs at https://dev.mysql.com/doc/refman/8.0/en/string-functions.html#function_substring

> A value of 0 for pos returns an empty string.

**Comment by camsaul** [comment]
@robdaemon I think we should work around the MySQL behavior so this behaves in a predictable way across different databases, If we wanted to document `substring()` somewhere, ideally we wouldn't have to add a note saying MySQL does something different than what you'd expect. It's easy enough for us to generate SQL that matches what you meant and work around MySQL's weird behavior.

MBQL `substring` is not meant to be a 1:1 map to the underlying SQL function, but rather a generalized string operation that we can compile to the appropriate native code. Exposing weird MySQL behavior seems to me like our abstraction is leaking details the user shouldn't have to care about.

@salsakran @flamber @sb interested in your opinions on the UX here -- I think we should reopen this and tweak the SQL we generate for MySQL. But maybe there's an argument for more closely following the behavior of the underlying DB?

**Comment by sbelak** [comment]
We should definitely strive for MBQL to work consistently across different DBs where possible. As far as I can tell `substring` in all the DBs indexes from 1. The ideal solution here would be to add appropriate validation on the BE (trivial) and FE (somewhat more work as we have only limited type information there -- perhaps another reason to step it up with sharing the mbql spec between FE & BE) and show the user an error.

**Comment by flamber** [comment]
It would be great if MBQL standardized this. Otherwise it needs to be documented (and perhaps a warning too) and helper shouldn't show it as the example. Reopening.

**Comment by camsaul** [comment]
Yeah, I think it makes sense to validate this in the backend. The current behavior seems like a bug, at least an error message would make it clear that what you're trying to do isn't going to work the way you think it is.

**Comment by robdaemon** [comment]
Makes sense. The only concern I would have is potential confusion with a query behaving differently in native SQL vs the question builder on MySQL here, but we may already have other cases of that I don't know about.

**Comment by sbelak** [comment]
I'll add the constraint to the MBQL schema.

**Comment by camsaul** [comment]
@robdaemon, you are right that it can be confusing when a pseudo-function in the expression editor named `substring()` doesn't work the same way as `substring()` in a native MySQL SQL query. 

There are two ways these sorts of expression pseudo-functions could work:

1.  Match behavior of underlying function exactly, even if it's different between different databases
2.  Provide the same behavior across every database we support

I think @sbelak and I are both on the same page that behavior number 2 is the one we want here. We should figure out how to communicate this clearly to users so someone who expects it to work like number 1 doesn't get confused and think Metabase is buggy when it doesn't work that way.

@kdoh @tlrobinson I'm interested in your thoughts on how to address this. Would it be feasible to pop up a little docstring explaining the semantics of `substring()` or other operations in the expression editor? I think that would be the ideal solution. Failing that, we should at least document this behavior

**Comment by dpsutton** [comment]
Leaving a breadcrumb here: we use substrings for fingerprinting to bound the size of strings that come back https://github.com/metabase/metabase/blob/master/src/metabase/db/metadata_queries.clj#L119
It is currently hardcoded to be `[:substring [:field (u/the-id field) nil] 1 truncation-size]` so this issue doesn't affect us. But if there is a smarter way to do it that would be great. Getting this wrong could cause us to not identify things based on those strings: json, email, us states, url.

**Comment by camsaul** [comment]
Let's investigate this and see if we can come up with a reasonable fix here if a reasonable amount of time

**Comment by john-metabase** [comment]
I opened a quick PR yesterday with a fix for MySQL, and I added a test to `metabase.query-processor-test.string-extracts-test/test-substring` with a `0` index, and CI showed failures for 7 additional drivers, so I closed that PR and coming back here to dig a little more.

We know that `1` works for all drivers, so I'm wondering if we should put a lower bound of 1 on the argument, or just always convert a start index of `0` to `1`? Some databases support negative indexes. Do we want to explicitly or implicitly support that through MBQL?

**Comment by john-metabase** [comment]
This test case:

```clojure
   (is (= "Red" (test-string-extract [:substring [:field (data/id :venues :name) nil] 0 3])))
 ```
 
The outputs for various drivers that failed:

```clojure
:presto    ""
:athena    ""
:postgres  "Re"
:redshift  "Re"
:sqlite    "Re"
:sqlserver "Re"
:vertica   "Re"
```

Adding code for `:substring` to all of those drivers individually feels like the wrong answer. Maybe this is best accomplished with `database-supports?`?
